src/packet/buffer_tx.rs

```diff
@@ -165,6 +203,80 @@ impl PacketizingBuffer {
     }
 }
 
+// Total queue time in buffer. This lovely drawing explains how to add more time.
+//
+// -time--------------------------------------------------------->
+//
+// +--------------+
+// |              |
+// +--------------+
+//      +---------+                          Already
+//      |         |                           queued
+//      +---------+                         durations
+//          +-----+
+//          |     |
+//          +-----+
+//                       +-+
+//                       | |         <-----  Add next
+//                       +-+                  packet
+//
+//
+//
+// +--------------+--------+
+// |              |@@@@@@@@|
+// +--------------+--------+
+//      +---------+--------+                 The @ is
+//      |         |@@@@@@@@|                  what's
+//      +---------+--------+                  added
+//          +-----+--------+
+//          |     |@@@@@@@@|
+//          +-----+--------+
+//                       +-+
+//                       |@|
+//                       +-+
+#[derive(Debug, Default)]
+struct TotalQueue {
+    unsent_count: usize,
+    unsent_size: usize,
+    last: Option<Instant>,
+    queue_time: Duration,
+}
+
+impl TotalQueue {
+    fn move_time_forward(&mut self, now: Instant) {
+        if let Some(last) = self.last {
+            assert!(self.unsent_count > 0);
+            assert!(self.unsent_size > 0);
+            let from_last = now - last;
+            self.queue_time += from_last * (self.unsent_count as u32);
+            self.last = Some(now);
+        } else {
+            assert!(self.unsent_count == 0);
+            assert!(self.unsent_size == 0);
+            assert!(self.queue_time == Duration::ZERO);
+        }
+    }
+
+    fn increase(&mut self, now: Instant, queue_time: Duration, size: usize) {
+        self.move_time_forward(now);
+        self.unsent_count += 1;
+        self.unsent_size += size;
+        self.queue_time += queue_time;
+        self.last = Some(now);
+    }
+
+    fn decrease(&mut self, size: usize, queue_time: Duration) {
+        self.unsent_count -= 1;
+        self.unsent_size -= size;
+        self.queue_time -= queue_time;
+        if self.unsent_count == 0 {
+            self.unsent_size = 0;
```

Comment:
This ought to be asserts, no? If the code is correct `self.unsent_size` and `self.queue_time` ought to have zero values here already, only `self.last` needs to actually be reset

Reply:
true that